addons/account/models/account_move_line.py

```diff
@@ -912,6 +912,7 @@ def _compute_tax_key(self):
                     'partner_id': line.partner_id.id,
                     'move_id': line.move_id.id,
                     'display_type': line.display_type,
+                    'tax_base_amount': line.tax_base_amount,
```

Comment:
This is part of the values, not of the keys
https://github.com/odoo/odoo/blob/248b8650ab63aeaaac162c12b3184bc96fc70080/addons/account/models/account_move_line.py#L965

Reply:
Beside `amount_currency`, `balance`, `id` and `tax_base_amount` the tax lines are identical.

What if we add support for multirecord to `existing` method https://github.com/odoo/odoo/commit/ff767da344c3735b1d5134f4eb088ae43da5dfed ?

Reply:
Why are they not merged together then? That's the purpose of this key dict.

Reply:
Because, since 932d0382b5a4a5d0d91dd532940bc2ed8d0c13da, in the key dict we have the `display_type`, which will make the difference between the 'product' tax line and 'epd' tax line

<details>
  <summary>The 'faulty' needed after</summary>
  
1 and 3 share the same key
```
 {
1) {'tax_repartition_line_id': 11902, 'group_tax_id': False, 'account_id': 23564, 'currency_id': 2, 'analytic_distribution': False, 'tax_ids': [(6, 0, [])], 'tax_tag_ids': [(6, 0, [])], 'partner_id': 2000, 'move_id': 3995, 'display_type': 'product'}: 
    {'name': 'Tax 15%', 'balance': -15.0, 'amount_currency': -15.0, 'tax_base_amount': 100.0},
2) {'id': 10802}: {'tax_tag_ids': [(6, 0, [])]},
3) {'tax_repartition_line_id': 11902, 'group_tax_id': False, 'account_id': 23564, 'currency_id': 2, 'analytic_distribution': False, 'tax_ids': [(6, 0, [])], 'tax_tag_ids': [(6, 0, [])], 'partner_id': 2000, 'move_id': 3995, 'display_type': 'epd'}:
     {'name': 'Tax 15%', 'balance': 1.5, 'amount_currency': 1.5, 'tax_base_amount': -10.0},
 4) {'id': 10803}: {'tax_tag_ids': [(6, 0, [])]},
 5) {'id': 10804}: {'tax_tag_ids': [(6, 0, [])]}
}
```  
</details>

To workaround we can revert the fix of 932d0382b5a4a5d0d91dd532940bc2ed8d0c13da
 And somehow manage the duplicated key in `existing()`
```diff
diff --git a/addons/account/models/account_move.py b/addons/account/models/account_move.py
index 382220f0e610..9b680d071a2c 100644
--- a/addons/account/models/account_move.py
+++ b/addons/account/models/account_move.py
@@ -1986,7 +1986,7 @@ class AccountMove(models.Model):
         def existing():
             return {
                 line[existing_key_fname]: line
-                for line in container['records'].line_ids
+                for line in container['records'].line_ids.sorted(key=lambda l: l.display_type == line_type)
                 if line[existing_key_fname]
             }
         def needed():
```

I fear it's not reliable but it passes all tests and it would work:
- in the original issue of 932d0382b5a4a5d0d91dd532940bc2ed8d0c13da the `write` would be called on the actual tax line (because of the sorting `existing` will keep the tax line instead of the rounding tax line)
- once we remove `display_type` from the key, the current PR issue is fixed as well

Reply:
Ooof, I should have seen that sooner :weary: 

Well, there are two solutions:
* Manage multiple lines with the same key in `existing()`
* Something like this
```diff
diff --git a/addons/account/models/account_move_line.py b/addons/account/models/account_move_line.py
index cc9e7c8aa05..1982ecbd615 100644
--- a/addons/account/models/account_move_line.py
+++ b/addons/account/models/account_move_line.py
@@ -911,7 +911,7 @@ class AccountMoveLine(models.Model):
                     'tax_tag_ids': [(6, 0, line.tax_tag_ids.ids)],
                     'partner_id': line.partner_id.id,
                     'move_id': line.move_id.id,
-                    'display_type': line.display_type,
+                    'display_type': 'epd' if _('(Discount)') in line.name else 'product',
                 })
             else:
                 line.tax_key = frozendict({'id': line.id})
@@ -958,7 +958,7 @@ class AccountMoveLine(models.Model):
                     'move_id': line.move_id.id,
                     'display_type': line.display_type,
                 }): {
-                    'name': tax['name'],
+                    'name': tax['name'] + (' ' + _('(Discount)') if line.display_type == 'epd' else ''),
                     'balance': tax['amount'] / rate,
                     'amount_currency': tax['amount'],
                     'tax_base_amount': tax['base'] / rate * (-1 if line.tax_tag_invert else 1),
```

I don't think that the sorting workaround actually works

Reply:
> By managing multiple lines with same key we would have 1 tax entry, like it was before [932d038](https://github.com/odoo/odoo/commit/932d0382b5a4a5d0d91dd532940bc2ed8d0c13da)

No, you'd still have multiple lines, but they would all be deleted if the key doesn't exist anymore